crates/ra_ide_api/src/inlay_hints.rs

```diff
@@ -35,35 +40,62 @@ fn get_inlay_hints(
     node: &SyntaxNode,
 ) -> Option<Vec<InlayHint>> {
     visitor()
-        .visit(|let_statement: ast::LetStmt| {
+        .visit(|let_statement: LetStmt| {
             if let_statement.ascribed_type().is_some() {
                 return None;
             }
+            let pat = let_statement.pat()?;
             let analyzer = SourceAnalyzer::new(db, file_id, let_statement.syntax(), None);
-            Some(get_pat_hints(db, &analyzer, let_statement.pat()?, InlayKind::LetBindingType))
+            Some(get_pat_hints(db, &analyzer, pat, InlayKind::LetBindingType, false))
         })
-        .visit(|closure_parameter: ast::LambdaExpr| {
+        .visit(|closure_parameter: LambdaExpr| {
```

Comment:
Perhaps, instead of visiting every construction where a pattern might appear, we should just visit `BindPat` directly?

Reply:
Not sure which approach will be better though!

Reply:
Is there a way to get a `PatKind` of a parent node from the `BindPat`?
I know how to get the `SyntaxKind`, but I need more to determine whether to add a hint or not based on the data about the `Pat`: for instance, for `let` bindings that have an ascribed type we don't display a type hint.

Otherwise it seems like a nice idea, seems to be of more or less the same complexity to the current implementation.
But I think I've covered all the cases we need the type hints for, haven't I?
If so, it might be ok to leave the code as is and refactor it after the https://github.com/rust-analyzer/rust-analyzer/issues/1618 is done, since it will help a lot with detecting the leaf `BindPat`s.

Reply:
>Is there a way to get a PatKind of a parent node from the BindPat

```rust
pat.syntax().parent().and_then(ast::Pat::cast).map(|it| it.kind())
```

>If so, it might be ok to leave the code as is and refactor it after the #1618 is done, since it will help a lot with detecting the leaf BindPats.

Makes sense!

Reply:
> Makes sense!

@matklad , the time had come to continue the discussion :)
I'm trying to apply the 

> Perhaps, instead of visiting every construction where a pattern might appear, we should just visit BindPat directly?

suggestion as

```rust
ast::BindPat(it) => {
    let pat = dbg!(dbg!(it).pat())?;
    let ty = dbg!(analyzer.type_of_pat(db, &pat))?;
    if ty.is_unknown() {
        return None;
    }

    acc.push(
        InlayHint {
            range: pat.syntax().text_range(),
            kind: InlayKind::TypeHint,
            label: ty.display_truncated(db, max_inlay_hint_length).to_string().into(),
        }
    );
},
```

yet in every test the `pat` is `None`:

```
[crates/ra_ide/src/inlay_hints.rs:94] it = BindPat {
    syntax: BIND_PAT@[69; 71)
      NAME@[69; 71)
        IDENT@[69; 71) "zz"
    ,
}
[crates/ra_ide/src/inlay_hints.rs:94] dbg!(it).pat() = None
[crates/ra_ide/src/inlay_hints.rs:94] it = BindPat {
    syntax: BIND_PAT@[105; 111)
      NAME@[105; 111)
        IDENT@[105; 111) "zz_ref"
    ,
}
[crates/ra_ide/src/inlay_hints.rs:94] dbg!(it).pat() = None
```

Is there something that I'm missing?

Reply:
Oh, looks like `let pat = ast::Pat::from(it);` should be used instead of `let pat =it.pat()?;`
What does the latter method return then?

Reply:
It returns the pattern in `name @ pattern` syntax.

Reply:
Hm, interesting usecase for the type hints actually, thank you.